lib/src/revset.rs

```diff
@@ -1743,9 +1730,19 @@ pub fn evaluate_expression<'repo>(
             Ok(Box::new(UnionRevset { set1, set2 }))
         }
         RevsetExpression::Intersection(expression1, expression2) => {
-            let set1 = expression1.evaluate(repo, workspace_ctx)?;
-            let set2 = expression2.evaluate(repo, workspace_ctx)?;
-            Ok(Box::new(IntersectionRevset { set1, set2 }))
+            match expression2.as_ref() {
+                RevsetExpression::Filter(predicate) => Ok(Box::new(FilterRevset {
+                    candidates: expression1.evaluate(repo, workspace_ctx)?,
+                    predicate: build_predicate_fn(repo, predicate),
+                })),
```

Comment:
Ah, it wasn't clear from the description that you would make the evaluation of `Intersection` smarter, but that makes sense. Nice.